[YDS-#228] Component - Toast 구현

Gael-Android opened the pull request:
### Summary
- 기존 YdsScaffold가 Toast를 수용할 수 없음 => NewScaffold 구현
- Toast 구현
- ToastHost 구현

## Describe your changes

### Toast in NewScaffold
![Toast](https://github.com/yourssu/YDS-Android/assets/84790707/9c0c7a84-b431-4bae-b3db-b17168d76643)

### Toast 
![image](https://github.com/yourssu/YDS-Android/assets/84790707/750b47ab-a33a-4f16-af1d-4ce63d19b6e9)

## Issue
- #228

Gael-Android committed:
MyScaffold, Toast 초기 후구현

Gael-Android committed:
구현 완료

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/base/NewScaffold.kt:
```diff
@@ -0,0 +1,198 @@
+
+
+@Composable
+fun NewScaffold(
```
YdsScaffold로 통합해주시면 좋을 것 같아요

지금 YdsScaffold가 할 수 있는 일을 NewScaffold가 모두 할 수 있다면 굳이 둘을 나눠둘 필요는 없을 것 같아요!

Gael-Android replied:
기존 YdsScaffold 대체하겠습니다~

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/base/NewScaffold.kt:
```diff
@@ -0,0 +1,198 @@
+    }
+}
+
+@Preview
+@Composable
+fun NewScaffoldPreview() {
+    val toastHostState: ToastHostState = remember { ToastHostState() }
```
preview는 중간에 있으면 헷갈려서 파일 맨 아래에 두는게 좋을 것 같아요. 그리고 private으로 설정해주세요!

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastHost.kt:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Stable
+    private class ToastDataImpl(
```
클래스 밖으로 빼주시는게 보기 편할 것 같아요

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastHost.kt:
```diff
@@ -0,0 +1,205 @@
+}
+
+
+interface ToastData {
+    val message: String
+    val duration: ToastDuration
+
+    fun dismiss()
+}
+
+enum class ToastResult {
+    Dismissed,
+}
+
+enum class ToastDuration {
+    SHORT, LONG,
+}
```
인터페이스, enum은 파일 맨 상단에 두면 좋을 것 같아요

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastHost.kt:
```diff
@@ -0,0 +1,205 @@
+                val duration = if (isVisible) ToastFadeInMillis else ToastFadeOutMillis
+                val opacity = animatedOpacity(
+                    animation = tween(
+                        easing = CubicBezierEasing(0.25f, 0.1f, 0.25f, 1f),
```
YdsEasing 사용해주세요!(아마 똑같긴 할거예요)

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastHost.kt:
```diff
@@ -0,0 +1,205 @@
+    val transition: FadeInFadeOutTransition
+)
+
+private typealias FadeInFadeOutTransition = @Composable (content: @Composable () -> Unit) -> Unit
```
이런 정의도 파일 상단으로 빼주세욥

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Toast.kt:
```diff
@@ -0,0 +1,153 @@
+    contentColor: Color = YdsTheme.colors.textBright,
+    toastData: ToastData
+) {
+    Layout({
```
```
Layout(
    content = { ... }
)
```
이런 식으로 명시해주는게 좋을 것 같아요

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Toast.kt:
```diff
@@ -0,0 +1,153 @@
+
+@Preview(showBackground = false, showSystemUi = true)
+@Composable
+fun ToastPreview() {
```
preview는 private으로 해주세요! (혹시 다른 파일의 프리뷰도 private이 없다면 붙여주시면 감사하겠습니다)

private으로 하는 이유는 public이면 외부에서 호출할 수 있기 때문이에요. 큰 의미는 없지만 사용할 때 헷갈리니까요

Gael-Android replied:
생각보다 안 돼어 있는게 많네요.

cometj03 reviewed (commented):
수고하셨습니다!

아직 구체적인 구현을 모두 보진 못했고 일단 형식적인 것만 먼저 리뷰 드렸어요. 이거 먼저 수정 부탁드립니다

그리고 왜 이런 방식을 사용해야 하는지, 또는 어떤 코드인지에 대한 설명을 주석으로 달아주시면 좋을 것 같아요 (단순한 구현이 아니다보니 처음 보면 이해하는데 시간이 걸릴 것 같아서요)

Gael-Android committed:
리뷰 반영

- 기존 YdsScaffold 대체
- preview 맨 아래로, private 변경
- 기존 다른 파일들도 preview function은 전부 private하게 변경
- 인터페이스, enum 맨 위로
- YdsEasing 사용
- Toast Animation 파일 분리
- 기타 코드 정리

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/base/YdsScaffold.kt:
```diff
@@ -1,39 +1,201 @@
     backgroundColor: Color = YdsTheme.colors.bgNormal,
-    contentColor: Color = LocalContentColor.current,
-    content: @Composable () -> Unit
+    contentColor: Color = YdsTheme.colors.textPrimary,
```
contentColor 기본값을 바꾼 이유가 있을까요?

Gael-Android replied:
잘못 바꿨네요...

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/base/YdsScaffold.kt:
```diff
@@ -1,39 +1,201 @@
+}
+
+@Composable
+fun ScaffoldLayout(
```
얘는 아마 private이어야 할 것 같아요!

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Toast.kt:
```diff
@@ -0,0 +1,137 @@
+                    modifier = Modifier.align(Alignment.Center)
+                )
+            }
+        }) { measurables, constraints ->
+        require(measurables.size == 1) {
+            "text for Snackbar expected to have exactly only one child"
+        }
```
음 뭔가 들여쓰기가 잘못된 것 같아요

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Toast.kt:
```diff
@@ -0,0 +1,137 @@
+        require(firstBaseline != AlignmentLine.Unspecified) { "No baselines for text" }
+        require(lastBaseline != AlignmentLine.Unspecified) { "No baselines for text" }
+
+        val minHeight =
+            if (firstBaseline == lastBaseline) {
+                48.dp
+            } else {
+                68.dp
+            }
```
아하 이런 식으로 한 줄인지 여러 줄인지 판단하는거군요

근데 48과 68은 어디에서 얻은 값인가요?

Gael-Android replied:
일단 임의의 값으로 하긴했는데... Figma에 있는값으로 변경하겠습니다.

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastAnimation.kt:
```diff
@@ -0,0 +1,105 @@
+
+typealias FadeInFadeOutTransition = @Composable (content: @Composable () -> Unit) -> Unit
+
+class FadeInFadeOutState<T> {
+    // we use Any here as something which will not be equals to the real initial value
+    var current: Any? = Any()
+    var items = mutableListOf<FadeInFadeOutAnimationItem<T>>()
+    var scope: RecomposeScope? = null
+}
+
+data class FadeInFadeOutAnimationItem<T>(
+    val key: T,
+    val transition: FadeInFadeOutTransition
+)
```
제가 이해하기론 토스트가 끝나기 전에 다른 토스트를 띄웠을 때를 염두에 둔 구현 방식인 것 같은데 맞을까요?

주석으로 설명해주시면 좋을 것 같습니다!

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastAnimation.kt:
```diff
@@ -0,0 +1,105 @@
+import com.yourssu.design.system.compose.rule.Duration
+import com.yourssu.design.system.compose.rule.YdsInAndOutEasing
+
+typealias FadeInFadeOutTransition = @Composable (content: @Composable () -> Unit) -> Unit
```
이 transition은 어디에서 사용되는 건가요?

Gael-Android committed:
리뷰 반영

- 코드수정
- 네이밍 변경
- 주석 작성중.
- 이해완료

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Toast.kt:
```diff
@@ -60,11 +62,13 @@ fun Toast(
             } else {
-                68.dp
+                ToastTwoLineMinHeight
             }
```
음 근데 최소 높이를 따로 설정해주는 이유를 잘 모르겠습니다.. 이미 토스트 content를 그릴 때 높이가 정해지는 것 아닌가요?

Gael-Android replied:
그러게요?

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastAnimation.kt:
```diff
@@ -3,89 +3,107 @@ package com.yourssu.design.system.compose.foundation
+ * @param toast 토스트 컴포저블
+ */
 @Composable
 fun FadeInFadeOutWithScale(
```
스케일이랑은 관련이 없는 것 같은데 with scale이라는 접미사가 붙은 이유가 무엇인가요?

Gael-Android replied:
이것도 예전에 있었던 기능 때문인데,, 미처 신경을 못썼네요.. 수정하겠습니다,.

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastAnimation.kt:
```diff
@@ -3,89 +3,107 @@ package com.yourssu.design.system.compose.foundation
-    val key: T,
-    val transition: FadeInFadeOutTransition
-)
+typealias opacityTransition = @Composable (toast: @Composable () -> Unit) -> Unit
```
타입 이름은 대문자로 시작해주세요

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/foundation/ToastAnimation.kt:
```diff
@@ -3,89 +3,107 @@ package com.yourssu.design.system.compose.foundation
-    var items = mutableListOf<FadeInFadeOutAnimationItem<T>>()
-    var scope: RecomposeScope? = null
-}
+fun Any?.helpCode() = System.identityHashCode(this) % 1000
```
이건 어디에서 쓰이는건가요?

Gael-Android replied:
이건 디버그 용도였는데 제거를 안했네요;

cometj03 replied:
이거 아직 그대로 있네요

Gael-Android replied:
헉

cometj03 reviewed (commented):
지금 보니 ToastAnimation과 ToastHost가 foundation에 들어가 있는 이유가 무엇인가요?

특별한 이유가 없다면 component 패키지 내부로 옮겨주시면 감사하겠습니다. toast 패키지를 내부에 추가로 만들어도 좋구요

cometj03 commented:
아! 그리고 버전도 하나 올려주셔야 라이브러리 배포할 때 반영되니까 버전 올리는 것도 잊지 말아주세요!

Gael-Android commented:
혼자 만드니까 멍청해지네요;; 리뷰가 반드시 필요하겠네요.....

Gael-Android committed:
리뷰 반영

- 패키지 변경
- 최소높이 제거
- 네이밍 변경
- 주석 작성

Gael-Android committed:
Merge branch 'develop' into feature/gael/toast

Gael-Android commented:
FadeInFadeOut에 대해서 주석을 작성했는데 이게 좀 복잡해서 제가 주석을 잘 작성했는지가 의심이 되네요.. Comet님 보시고 알려주세요!!!

Gael-Android commented:
이건 보실때 참고하라고 올려두겠습니다. 내부 상태를 추적하기가 힘들어서 도움이 될 거 같아요

- 에니메이션이 끝날때까지 기다리면서 눌렀을떄
![ToastOneByOne](https://github.com/yourssu/YDS-Android/assets/84790707/e94a24f0-7f4c-4ad7-ac49-42425b41dfdd)
![Pasted image 20231229002055](https://github.com/yourssu/YDS-Android/assets/84790707/19af6062-f685-43a3-87d4-202fbbf31828)

- 에니메이션이 끝나기 전에 여러번 눌렀을때
![ToastRapidThree](https://github.com/yourssu/YDS-Android/assets/84790707/15b3fd02-48f2-4d43-b127-c08e12d7fca0)
![Pasted image 20231229002431](https://github.com/yourssu/YDS-Android/assets/84790707/604da7c3-3086-44d5-b8ca-cb708a51b373)

cometj03 commented:
넵 감사합니다! 내일 시간 될 때 찬찬히 볼게요

cometj03 commented:
그리고 리뷰 반영 커밋은 개수가 많아지더라도 하나씩 나눠서 해주시는 게 확인하기 편할 것 같아요!
(빠뜨린 부분이 생길 수 있어서요. 확인 부탁드립니다)

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/toast/Toast.kt:
```diff
@@ -57,20 +50,8 @@ fun Toast(
-        val containerHeight = max(minHeight.roundToPx(), textPlaceable.height)
         layout(constraints.maxWidth, containerHeight) {
             val textPlaceY = (containerHeight - textPlaceable.height) / 2
             textPlaceable.placeRelative(0, textPlaceY)
```
더이상 Layout으로 위치를 지정해줄 필요 없이 그냥 가운데 정렬만 하면 될 것 같은데 확인 부탁드립니다

(근데 토스트 명세에 텍스트가 한 줄이면 가운데 정렬, 여러 줄이면 왼쪽 정렬해야 하는데 이 부분을 처리하는 로직은 어디에 있는건가요?)

Gael-Android replied:
이게 정말 이상한데, 
//            modifier = Modifier.align(
//                Alignment.Center
//            )
를 하기만 해도 텍스트가 한줄일때는 중앙정렬이 되고, 2줄이상 넘어갈때 좌측정렬이 자동으로 됩니다.

원래 방법으로는,
//            onTextLayout = {
//                lineCount = it.lineCount
//            },       
를 YdsText에 하고, 
//                if (lineCount == 1)
//                    Alignment.Center
//                else
//                    Alignment.CenterStart
하면 정해진 로직대로 구현을 하는 방법입니다. 하지만 첫번째 방법이 이해할 수 없지만 작동하니 
이건 어떻게 해야할지 잘 모르겠습니다. (이해도 안되고요...)

Gael-Android replied:
그래도 일단 기획이 코드에 보여야하니까 아래의 방법으로 하겠습니다.

cometj03 replied:
아하 그러게요 희한하네요

텍스트 정렬하는 방법을 찾아보니 보통 modifier가 아니라 textAlign 속성을 지정해주는 방식으로 하더라구요 (https://kotlinworld.com/219)
Modifier.align이랑은 무슨 차이가 있을까요..?

cometj03 replied:
내부 구현을 둘러보니까 일단 align 함수는 Box나 Column, Row 스코프 내에서만 사용할 수 있는 함수예요. 컨테이너 내부의 각 요소에 대해 어떻게 둘거냐를 지정하는거죠. 그래서 텍스트의 속성이라고 보긴 어려울 것 같아요.

그리고 Modifier.align(Center)를 하든 안 하든 텍스트 정렬은 `TextAlign.Start`(좌측 정렬)가 기본이에요

```kotlin
// 참고
// androidx.compose.ui.text.ParagraphTextStyle

internal val textAlignOrDefault: TextAlign = textAlign ?: TextAlign.Start
```

근데 왜 말씀하신 것처럼 한 줄일 때는 중앙 정렬이고 여러 줄일 때는 좌측 정렬인 것처럼 보일까요? 그림으로 직관적으로 보여드리자면
![image](https://github.com/yourssu/YDS-Android/assets/48352078/7a69ff38-31ec-4e0c-ba6a-44581f3361e0)

이렇게 그냥 텍스트가 짧으면 차지하는 영역이 작으니까 Box의 Center 속성 때문에 가운데에 있는 것처럼 보이는 거예요.

cometj03 replied:
위 링크의 textAlign 속성은 머티리얼 Text에만 있는거더라구요. 그래서 의도대로 하기 위해서는 YdsText를 수정해야 하는데 글로 쓰기는 길어지니까 이 PR 머지되고 나서 제가 직접 수정할게요. 지금은 그냥 이 상태로 두셔도 괜찮습니다!

Gael-Android replied:
> 내부 구현을 둘러보니까 일단 align 함수는 Box나 Column, Row 스코프 내에서만 사용할 수 있는 함수예요. 컨테이너 내부의 각 요소에 대해 어떻게 둘거냐를 지정하는거죠. 그래서 텍스트의 속성이라고 보긴 어려울 것 같아요.
> 
> 
> 
> 그리고 Modifier.align(Center)를 하든 안 하든 텍스트 정렬은 `TextAlign.Start`(좌측 정렬)가 기본이에요
> 
> 
> 
> ```kotlin
> 
> // 참고
> 
> // androidx.compose.ui.text.ParagraphTextStyle
> 
> 
> 
> internal val textAlignOrDefault: TextAlign = textAlign ?: TextAlign.Start
> 
> ```
> 
> 
> 
> 근데 왜 말씀하신 것처럼 한 줄일 때는 중앙 정렬이고 여러 줄일 때는 좌측 정렬인 것처럼 보일까요? 그림으로 직관적으로 보여드리자면
> 
> ![image](https://github.com/yourssu/YDS-Android/assets/48352078/7a69ff38-31ec-4e0c-ba6a-44581f3361e0)
> 
> 
> 
> 이렇게 그냥 텍스트가 짧으면 차지하는 영역이 작으니까 Box의 Center 속성 때문에 가운데에 있는 것처럼 보이는 거예요.

정말 그렇네요!! 저 이미지가 맞는거 같습니다.

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/base/YdsScaffold.kt:
```diff
@@ -1,39 +1,202 @@
+@Preview
+@Composable
+private fun YdsScaffoldPreview() {
+    val toastHostState: ToastHostState = remember { ToastHostState() }
```
사용할 때 remember를 빠뜨릴 가능성이 있으니까 `rememberYdsToastHostState()` 같은 함수를 제공해주는 것도 좋을 것 같아요

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/toast/ToastHost.kt:
```diff
@@ -0,0 +1,101 @@
+    Dismissed,
+}
+
+enum class ToastDuration {
+    SHORT, LONG,
+}
```
각 enum에 프로퍼티를 정해도 좋을 것 같아요
`SHORT(1500L)` 처럼 정의하고 사용할 땐 `delay(duration.value)`처럼 사용할 수 있게요

Gael-Android replied:
좋은 생각입니다!

Gael-Android committed:
디버그 용 함수 제거

Gael-Android committed:
rememberToastHostState 생성

Gael-Android committed:
ToastDuration Enum애 시간 정의

Gael-Android committed:
텍스트 가운데 정렬 로직 추가

haeti-dev commented on compose/src/main/java/com/yourssu/design/system/compose/component/toast/ToastAnimation.kt:
```diff
@@ -0,0 +1,124 @@
+        scheduledToastData = newToastData // 앞으로 나타날 토스트에 새로운 토스트로 업데이트
+        val toastDataList = toastTransitions.map {
+            it.toastData
+        }.toMutableList() // 현재 나타난 토스트의 정보로 초기회
```
아주 사소한거지만 주석에 오타가 있네요!
초기회 -> 초기화

haeti-dev reviewed (commented):
컴포즈 초짜라 리뷰달건 없지만, 덕분에 많이 배우고 있습니다!
친절한 주석도 감사합니다 :)

Gael-Android commented:
> 컴포즈 초짜라 리뷰달건 없지만, 덕분에 많이 배우고 있습니다!
> 
> 친절한 주석도 감사합니다 :)

저도 잘 모릅니다. 부담가지시지 마시고 정말 무엇이든 물어봐주세요! 사소한 것이라도요.

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/base/YdsScaffold.kt:
```diff
@@ -51,6 +51,9 @@ fun YdsScaffold(
 }
 
+@Composable
+fun rememberToastHostState(): ToastHostState = remember { ToastHostState() }
```
아래 preview에서도 해당 함수 사용하는 것으로 수정해주시면 좋을 것 같아요
그리고 이 함수 선언은 ToastHost 파일에 있는 게 더 자연스러울 것 같네요!

Gael-Android committed:
rememberToastHost -> ToastHost.kt로 이동

Gael-Android committed:
오타 수정
